### DISCUSSION TITLE
Image/Page Matching Issue when using PDF file converted by LibreOffice

### DISCUSSION THREAD

**Discussion by ChristophGmeiner** [py-pdf/pypdf#2823] [Show and tell]
See all details and code here on ther LibreOffice site: https://ask.libreoffice.org/t/pdf-encoding-issues-with-page-image-matching-when-converting-docx-to-pdf/110150/3

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
macOS-14.5-arm64-arm-64bit

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==4.3.1, crypt_provider=('local_crypt_fallback', '0.0.0'), PIL=10.4.0

```

**Comment by pubpub-zz** [comment]
please provide the generated pdf

**Comment by ChristophGmeiner** [comment]
[test04_libreoffice_exportwriter.pdf](https://github.com/user-attachments/files/16817695/test04_libreoffice_exportwriter.pdf)

**Comment by pubpub-zz** [comment]
There is no bug in pypdf:
libreoffice attaches all the images to all the pages
![image](https://github.com/user-attachments/assets/405ee978-4017-4faf-93a3-38d5030f1db9)

it is only within the contents that only some images are called.
If you want only theimages used your code should look like this:
```
for p in pdf_reader.pages:
                print(p.page_number)
                print([img.name for img in p.images if img.name.split('.')[0] in [op[0][0][1:] for op in p.get_contents().operations if op[1]==b"Do"]])
```
img.name contains the key but has no "/" and has the extension so the code is a little tricky

**Comment by mikekaganski** [comment]
@pubpub-zz is the code that you provided the universal code for images used on specific pages; or is it only for the PDFs generated by LibreOffice? Is the behavior of LibreOffice a bug, or is it a valid variant?

**Comment by stefan6419846** [comment]
The provided code should be mostly universal, as it basically checks which image objects actually are referenced in the page itself on plain PDF operator level.

I am not sure whether the standard says anything about this, so you might want to have a look at it yourself, but from my experience the behavior of LibreOffice in this case is something I usually do not see from other PDF generators.

**Comment by pubpub-zz** [comment]
the standard provides "grammar" not "stories". This approach is compliant and i've seen it many times. also if something is to be "blamed" it is libreoffice, not pypdf.
I convert this as a discussion for knowledge

**Comment by baileymorton989** [comment]
I am experiencing the same issue. Has anyone looked into this?

@ChristophGmeiner were you able to solve this?

**Reply by ChristophGmeiner** [reply]
The provided code snippet above solved the issue.

**Reply by baileymorton989** [reply]
@ChristophGmeiner do you mean the following:

`for p in pdf_reader.pages:
                print(p.page_number)
                print([img.name for img in p.images if img.name.split('.')[0] in [op[0][0][1:] for op in p.get_contents().operations if op[1]==b"Do"]])`

**Reply by baileymorton989** [reply]
@ChristophGmeiner And that extracts the images in the correct order?

**Reply by ChristophGmeiner** [reply]
Yes. But you still need to perform some deduplication afterwards

**Reply by baileymorton989** [reply]
Perfect! Yep, been doing my own de-duping so should be good there. Was there any de-duping you found particularly useful?

**Reply by pubpub-zz** [reply]
> @ChristophGmeiner And that extracts the images in the correct order?

the image list is created in the same order as they are in the pdf structure, this does not reflect the order where they are identified within contents, and this order may be different from the visible order on the page (images are floating on the page and do not use the same coordinates as text)

do not expect pypdf to do more than I can do
